components/core/src/clp/FileDescriptor.cpp

```diff
@@ -0,0 +1,57 @@
+#include "FileDescriptor.hpp"
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <cerrno>
+#include <cstddef>
+#include <string_view>
+
+#include "ErrorCode.hpp"
+#include "type_utils.hpp"
+
+namespace clp {
+FileDescriptor::FileDescriptor(
+        std::string_view path,
+        OpenMode open_mode,
+        CloseFailureCallback close_failure_callback
+)
+        : m_fd{(0 != (enum_to_underlying_type(open_mode) & O_CREAT))
+                       ? open(path.data(), enum_to_underlying_type(open_mode), S_IRWXU)
```

Comment:
* `S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH`?
* Also, can we do this in the body of the method instead?
* Can we add a comment for the writable case saying something like:
`// Enable writing for the owner and reading for everyone. Callers can change the created file's permissions as necessary.`